design/feature-flags.md

```diff
@@ -0,0 +1,71 @@
+# Feature Flags
+
+Status: Draft
+
+Some features may take a while to get fully implemented, and we don't necessarily want to have long-lived feature branches
+A simple feature flag implementation allows code to be merged into master, but not used unless a flag is set.
+
+## Goals
+
+- Allow unfinished features to be present in Velero releases, but only enabled when the associated flag is set.
+
+## Non Goals
+
+- A robust feature flag library.
+
+## Background
+
+When considering the [CSI integration work](https://github.com/heptio/velero/pull/1661), the timelines involved presented a problem in balancing a release and longer-running feature work.
+A simple implementation of feature flags can help protect unfinished code while allowing the rest of the changes to ship.
+
+## High-Level Design
+
+A new command line flag, `--features` will be added to the root `velero` command.
+
+`--features` will accept a comma-separated list of features, such as `--features EnableCSI,Replication`.
+Each feature listed will correspond to a key in a map in `pkg/features/features.go` defining whether a feature should be enabled.
+
+Any code implementing the feature would then import the map and look up the key's value.
+
+For the Velero client, a `features` key can be added to the `config.json` file for more convenient client invocations.
+
+## Detailed Design
+
+A new `features` package will be introduced with these basic structs:
+
+```go
+type FeatureFlagSet struct {
+    flags map[string]bool
```

Comment:
Guess this could also be a `sets.String` type, unless we want to keep the option of being able to disable a flag?

Reply:
Oh, yeah, that's a good idea.